test/dtypes.c

```diff
@@ -3788,6 +3819,109 @@ test_compound_18(void)
     return 1;
 } /* end test_compound_18() */
 
+/*-------------------------------------------------------------------------
+ * Function:    test_user_compound_conversion
+ *
+ * Purpose:     Tests that library correctly handles a user-defined
+ *              conversion function between two compound types.
+ *
+ * Return:      Success:        0
+ *              Failure:        number of errors
+ *
+ *-------------------------------------------------------------------------
+ */
+static int
+test_user_compound_conversion(void)
+{
+    hid_t             file_id = H5I_INVALID_HID;
+    hid_t             dset_id = H5I_INVALID_HID;
+    hid_t             src = H5I_INVALID_HID, dst = H5I_INVALID_HID;
+    hid_t             space_id = H5I_INVALID_HID;
+    struct src_cmpd_t buf[]    = {{1, 1.0}};
+
+    hsize_t dim = 1;
+    char    filename[1024];
+
+    TESTING("compound conversion via user conversion callback");
+
+    /* Create the source compound datatype */
+    if ((src = H5Tcreate(H5T_COMPOUND, sizeof(struct src_cmpd_t))) < 0) {
+        FAIL_PUTS_ERROR("couldn't create compound datatype");
+    }
+
+    if (H5Tinsert(src, "a", HOFFSET(struct src_cmpd_t, a), H5T_NATIVE_UINT32) < 0) {
+        FAIL_PUTS_ERROR("couldn't insert field into compound datatype");
+    }
+
+    if (H5Tinsert(src, "b", HOFFSET(struct src_cmpd_t, b), H5T_NATIVE_FLOAT) < 0) {
+        FAIL_PUTS_ERROR("couldn't insert field into compound datatype");
+    }
+
+    /* Create the destination compound datatype */
+    if ((dst = H5Tcreate(H5T_COMPOUND, sizeof(struct dst_cmpd_t))) < 0) {
+        FAIL_PUTS_ERROR("couldn't create compound datatype");
+    }
+
+    if (H5Tinsert(dst, "b", HOFFSET(struct dst_cmpd_t, b), H5T_IEEE_F32LE) < 0) {
+        FAIL_PUTS_ERROR("couldn't insert field into compound datatype");
+    }
+
+    if (H5Tregister(H5T_PERS_SOFT, "src_cmpd_t->dst_cmpd_t", src, dst, &user_compound_convert) < 0) {
+        FAIL_PUTS_ERROR("couldn't register compound conversion function");
+    }
+
+    /* Create File */
+    h5_fixname(FILENAME[0], H5P_DEFAULT, filename, sizeof filename);
+    if ((file_id = H5Fcreate(filename, H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT)) < 0) {
+        FAIL_PUTS_ERROR("couldn't create file");
+    }
+
+    /* Create a dataspace to use */
+    space_id = H5Screate_simple(1, &dim, NULL);
+    assert(space_id > 0);
+
+    /* Create a dataset with the destination compound datatype */
+    if ((dset_id = H5Dcreate2(file_id, "dataset", dst, space_id, H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT)) <
+        0) {
+        FAIL_PUTS_ERROR("couldn't create dataset with compound datatype");
+    }
+
+    if (H5Dwrite(dset_id, src, space_id, H5S_ALL, H5P_DEFAULT, buf) < 0) {
+        FAIL_PUTS_ERROR("couldn't write to dataset with user compound conversion function");
```

Comment:
These should probably all just be TEST_ERROR. FAIL_STACK_ERROR is rarely needed and FAIL_PUTS_ERROR is usually only used when erroring on non-HDF5 errors. The convention is also to not use brackets around the error lines.